### PR TITLE
feat(aws_s3 sink): add option to use virtual addressing

### PR DESCRIPTION
**Pull request by sam6258**
## Summary

The following PR adds an option to enable virtual host style addressing to an s3 sink. The `force_path_style` option defaults to `true` for backwards compatibility and may eventually be desired to default to `false` as path style addressing will eventually be removed from aws s3.

This is my first Rust and first Vector PR, so please correct me on any contribution guidelines, style semantics, and code patterns!

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?

I pointed an aws s3 sink at my object storage service and validated both path style and vhost style addressing works and also has the correct `authority` in access logs.

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- [x] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
Thanks for this contribution @sam6258 !

Do you think it'd be possible to add an integration test that uses the virtual addressing style? It looks like [localstack supports it](https://hashnode.localstack.cloud/efficient-localstack-s3-endpoint-configuration).

**Comment by jszwedko** [comment]
Integration tests are defined here: https://github.com/vectordotdev/vector/blob/master/src/sinks/aws_s3/integration_tests.rs

File: website/cue/reference/components/sinks/base/aws_s3.cue
Comment (neko-dd) [inline website/cue/reference/components/sinks/base/aws_s3.cue:622]:
```suggestion
			This controls if the bucket name is in the hostname or is part of the URL.
```

**Review by neko-dd — APPROVED** [review]

**Comment by sam6258** [comment]
> Do you think it'd be possible to add an integration test that uses the virtual addressing style? 

@jszwedko working on this! My editor didn't let me know I broke integration tests in the call to `create_service` too. Is it possible to enable tests to run on this PR? I'm working on getting my dev env setup to run integration tests in the meantime.

**Comment by pront** [comment]
> > Do you think it'd be possible to add an integration test that uses the virtual addressing style?
> 
> @jszwedko working on this! My editor didn't let me know I broke integration tests in the call to `create_service` too. Is it possible to enable tests to run on this PR? I'm working on getting my dev env setup to run integration tests in the meantime.

I enabled the CI checks (because it requires approval by maintainers). 

Local iterations should be much faster. This might be helpful: https://github.com/vectordotdev/vector/blob/master/scripts/integration/README.md

**Comment by sam6258** [comment]
Adapted the integration tests for the new function call structure and added `s3_healthchecks_vhost`, but I'm getting a lot of  test timeouts locally for some reason. Will dig into a bit, since I can't run all the tests I need, but I think the current commit should get us what we want.

File: src/sinks/aws_cloudwatch_logs/integration_tests.rs
Comment (jszwedko) [inline src/sinks/aws_cloudwatch_logs/integration_tests.rs:470]: I think this means we will _only_ test with virtual path style addressing. Could we make sure we have at least one test using each style?

Comment (sam6258) [reply src/sinks/aws_cloudwatch_logs/integration_tests.rs]: I got the integration tests working locally and pushed the changes. I had some incorrect values in the test client configs.

On a side note, I did not end up adding an integration test for virtual host style access. This proved more difficult than initially intended as it requires wildcard DNS which docker compose does not support easily. Localstack does support virtual host based lookups and I was able to set up the sdk to do so, but it fails on resolution of the actual bucket. I think wildcard DNS might be out of the scope of this PR.

What are your thoughts here?

Comment (sam6258) [reply src/sinks/aws_cloudwatch_logs/integration_tests.rs]: Here is the branch I was playing with if interested https://github.com/sam6258/vector/tree/sam6258/vhost-buckets-test

**Comment by sam6258** [comment]
@pront I fixed the CI check that was failing if you can kick off again, thanks!

**Comment by pront** [comment]
> @pront I fixed the CI check that was failing if you can kick off again, thanks!

Is PR now ready for a final review? There has been difficulty with getting an IT running?

**Comment by sam6258** [comment]
Yes, PR in my opinion is ready for final review. It passes existing tests, but I wasn't able to get a test to cover the new feature per this comment. https://github.com/vectordotdev/vector/pull/21999#discussion_r1881099659

**Comment by ADustyOldMuffin** [comment]
Is there anything else we need to do to get this merged?

**Comment by pront** [comment]
> Is there anything else we need to do to get this merged?

Leaving it this up to @jszwedko, due to this thread:
https://github.com/vectordotdev/vector/pull/21999#discussion_r1879732038

File: src/aws/mod.rs
Comment (jszwedko) [inline src/aws/mod.rs:129]: I'm not wild about having an S3 specific method on this generic trait.

What would think of, instead of adding this method here, adding a field to https://github.com/vectordotdev/vector/blob/475e93f182cc51e06ca74f8545a245203bcec1d0/src/common/s3.rs#L5 that controls the argument to `force_path_style`?

**Review by jszwedko — COMMENTED** [review]
Thanks @sam6258 ! I see that it'd be difficult to add an integration test due to the necessity of docker-compose supporting wildcard domains. In this case, I think we can just rely on the SDK to do the right thing.

I left one comment about changing the way the configuration is injected, but otherwise I think this looks good to go.

File: src/aws/mod.rs
Comment (sam6258) [reply src/aws/mod.rs:129]: Yeah, I wasn't crazy about this either, but lacking rust skills here didn't know how else to achieve this. Would that look something like this?

```
pub(crate) struct S3ClientBuilder;

// field here?

impl ClientBuilder for S3ClientBuilder {
```

Comment (sam6258) [reply src/aws/mod.rs]: @jszwedko latest commit should address this! Lemme know what you think. Thanks

File: src/common/s3.rs
Comment (jszwedko) [inline src/common/s3.rs:17]: I think you could avoid the `if let Some` if you just do: `builder = builder.set_force_path_style(self.force_path_style)`. The `set_*` variants on the AWS SDK take `Option`s.

**Review by jszwedko — COMMENTED** [review]
Thanks @sam6258 ! This looks much better. Just one last minor comment. Otherwise I think this is good to go.

File: src/common/s3.rs
Comment (sam6258) [reply src/common/s3.rs]: Done! I opted to use `force_path_style` so I could unwrap to `true` by default in case someone forgets to set the value in `S3ClientBuilder`. That way it maintains vector's previous behavior even in a programming error.

**Review by jszwedko — APPROVED** [review]
Nice work @sam6258 !

**Pull request merged**
